File: implement-shell-tools/wc/my_wc.js
const fs = require('fs');

const args = process.argv.slice(2);

let flagL = false;
let flagW = false;
let flagC = false;
let files = [];

for (let i = 0; i < args.length; i++) {
    if (args[i] === '-l') {
        flagL = true;
    } else if (args[i] === '-w') {
        flagW = true;
    } else if (args[i] === '-c') {
        flagC = true;
    } else {
        files.push(args[i]);
    }
}

// default: all flags on if none given
if (!flagL && !flagW && !flagC) {
    flagL = true;
    flagW = true;
    flagC = true;
}

// function to count words
function wordCount(str) {
    let parts = str.split(' ');
    let count = 0;
    for (let i = 0; i < parts.length; i++) {
        if (parts[i] !== '') {
            count++;
        }
    }
    return count;
}

let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;

files.forEach(file => {
    const content = fs.readFileSync(file, 'utf8');
    const rawBytes = fs.readFileSync(file).length;
    
    // count lines: split by newline
    const linesArray = content.split('\n');
    const lineCount = linesArray.length;
    
    // count words: need to split by spaces AND newlines
    let wordCountManual = 0;
    let inWord = false;
    for (let i = 0; i < content.length; i++) {
        let ch = content[i];
        if (ch === ' ' || ch === '\n' || ch === '\t') {
            inWord = false;
        } else {
            if (!inWord) {
                wordCountManual++;
                inWord = true;
            }
        }
    }
    
    totalLines += lineCount;
    totalWords += wordCountManual;
    totalBytes += rawBytes;
    
    let outputParts = [];
    if (flagL) outputParts.push(lineCount);
    if (flagW) outputParts.push(wordCountManual);
    if (flagC) outputParts.push(rawBytes);
    outputParts.push(file);
    console.log(outputParts.join(' '));
});

if (files.length > 1) {
    let totalParts = [];
    if (flagL) totalParts.push(totalLines);
    if (flagW) totalParts.push(totalWords);
    if (flagC) totalParts.push(totalBytes);
    totalParts.push('total');
    console.log(totalParts.join(' '));
}

File: implement-shell-tools/wc/my_wc.py
import sys

args = sys.argv[1:]

flagL = False
flagW = False
flagC = False
files = []

for arg in args:
    if arg == '-l':
        flagL = True
    elif arg == '-w':
        flagW = True
    elif arg == '-c':
        flagC = True
    else:
        files.append(arg)

# default to all flags if none specified
if not flagL and not flagW and not flagC:
    flagL = True
    flagW = True
    flagC = True

def wordCount(str):
    parts = str.split(' ')
    count = 0
    for i in range(len(parts)):
        if parts[i] != '':
            count = count + 1
    return count

totalLines = 0
totalWords = 0
totalBytes = 0

for file in files:
    f = open(file, 'r')
    content = f.read()
    f.close()
    
    f2 = open(file, 'rb')
    rawBytes = len(f2.read())
    f2.close()
    
    linesArray = content.split('\n')
    lineCount = len(linesArray)
    
    # count words by checking each character
    wordCountManual = 0
    inWord = False
    for ch in content:
        if ch == ' ' or ch == '\n' or ch == '\t':
            inWord = False
        else:
            if not inWord:
                wordCountManual = wordCountManual + 1
                inWord = True
    
    totalLines = totalLines + lineCount
    totalWords = totalWords + wordCountManual
    totalBytes = totalBytes + rawBytes
    
    outputParts = []
    if flagL:
        outputParts.append(str(lineCount))
    if flagW:
        outputParts.append(str(wordCountManual))
    if flagC:
        outputParts.append(str(rawBytes))
    outputParts.append(file)
    print(' '.join(outputParts))

# show total line if multiple files
if len(files) > 1:
    totalParts = []
    if flagL:
        totalParts.append(str(totalLines))
    if flagW:
        totalParts.append(str(totalWords))
    if flagC:
        totalParts.append(str(totalBytes))
    totalParts.append('total')
    print(' '.join(totalParts))